### DISCUSSION TITLE
Custom plugin development

### DISCUSSION THREAD

**Discussion by fr05t1k** [roadrunner-server/roadrunner#982] [Q&A] [answered]
Hey guys,

I'm looking for best practices for plugin development. Currently, the only thing that comes to my mind, is just to fork the whole roadrunner repo and add changes in a separate branch. 

I tried to implement my own "main" package with _rr_ command, but everything exists in _internal_ directory and I cannot import it into my package. Also, I cannot alter _container.Plugins()_ function without changing the file directly.

Can you suggest how to do it in the right way?

**Comment by rustatian** [comment, marked as answer]
Hey @fr05t1k. You don't need to fork any repository to add your custom plugin. A few days ago we introduced a new tool called [velox](https://github.com/roadrunner-server/velox).
So, to create your own plugin (only GitHub (private repos too) supported at the moment), you need to do the following:
1. If it's independent - create a repo, call the structure: [sample](https://github.com/roadrunner-server/grpc/blob/master/plugin.go#L29)
```go
type Plugin struct {}
```
2. If this is a middleware - have a look at the RR HTTP middleware API: [click](https://github.com/roadrunner-server/api/blob/master/plugins/middleware/interface.go), implement it on your structure.
3. If this is any plugin, which re-implements API - choose the proper interface and implement it.
4. Add your plugin to the velox configuration: [click](https://github.com/roadrunner-server/velox/blob/master/velox.toml).
```toml
you_super_plugin = { ref = "master", owner = "you", repository = "super_plugin" }
```
5. Build the RR: `vx build -c velox.toml -o /path/where/to/put/rr/`
6. Feel free to join our discord server: https://discord.gg/TFeEmCs

Next week I'll create a video tutorial about plugin development. It'll be on our YT channel: https://www.youtube.com/channel/UCHT6-FZGh5Y7JUGJyB8_M0Q/videos

**Reply by rustatian** [reply]
You can also have a look at the https://github.com/roadrunner-server/logger, `*/http`, `*/gzip` (middleware), `*/jobs`. All these repositories are plugins.

**Reply by rustatian** [reply]
All documentation is on the way 😃

**Reply by fr05t1k** [reply]
Wow! it looks promising. Thanks @rustatian it's exactly what I'm looking for!

**Reply by fr05t1k** [reply]
I hope there will be the ability to use standalone GitLab repositories :)

**Reply by rustatian** [reply]
> I hope there will be the ability to use standalone GitLab repositories :)

Yeah, sure. Just because GitLab has a complicated API (from my POV), I decided to implement GitHub API and then according to the demand move forward 😃
